### PR TITLE
Initial implementation of groupby().transform()

### PR DESCRIPTION
**Pull request by OliverHofkens**
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

File: dask/dataframe/tests/test_groupby.py [hunk 1]
```diff
@@ -844,6 +844,77 @@ def test_apply_shuffle():
...
+
+        assert_eq(
+            ddf.groupby(ddf["A"])[["B", "C"]].transform(lambda x: x.sum()),
+            pdf.groupby(pdf["A"])[["B", "C"]].transform(lambda x: x.sum()),
+        )
+
+        assert_eq(
+            ddf.groupby(ddf["A"] + 1)[["B", "C"]].transform(lambda x: x.sum()),
+            pdf.groupby(pdf["A"] + 1)[["B", "C"]].transform(lambda x: x.sum()),
+        )
```
[the remark this answers is not in the document]
Comment (OliverHofkens) [reply dask/dataframe/tests/test_groupby.py]: Great suggestion! I refactored the tests in ee119a88b8d0505d8edca970b821980b6aaee7b1 and combined them with the exact same tests for `groupby().apply()` through double parametrization as these are very similar functions.

File: dask/dataframe/tests/test_groupby.py [hunk 2]
```diff
@@ -1974,3 +2091,29 @@ def test_groupby_cov(columns):
         assert_eq(expected, result, check_index=False)
     else:
         assert_eq(expected, result)
+
+
+@pytest.mark.parametrize("transformation", [lambda x: x.sum(), np.sum, "sum"])
```
[the remark this answers is not in the document]
Comment (OliverHofkens) [reply dask/dataframe/tests/test_groupby.py]: I added 2 test cases for the second form in 45adbf73f7b2f09c7e7fe382c70aeb4f331dfafa and it seems to work fine!

File: dask/dataframe/groupby.py [hunk 3]
```diff
@@ -1511,6 +1527,136 @@ def apply(self, func, *args, **kwargs):
 
         return df5
 
+    @insert_meta_param_description(pad=12)
+    def transform(self, func, *args, **kwargs):
+        """ Parallel version of pandas GroupBy.transform
+
+        This mimics the pandas version except for the following:
+
+        1.  The user should provide output metadata.
```
[the remark this answers is not in the document]
Comment (OliverHofkens) [reply dask/dataframe/groupby.py]: I deleted that first point in 87697922addcabdea3bd05aa46ccca1a10d6cc93, as the common `meta` parameter documentation is present on both `apply` and `transform`. That seems sufficient to me.

**Comment by OliverHofkens** [comment]
I extracted the shuffling code that `apply` and `transform` share in 87697922addcabdea3bd05aa46ccca1a10d6cc93 to a `_shuffle` method. 

I'm gonna take a look at how transform interacts with custom `Aggregation()`s next, then merge master back in! Thanks for the feedback so far!

**Comment by TomAugspurger** [comment]
One quick question, since I haven't had time to look into the code yet, what is the type given to the user-defined function in

```python
df.groupby("column").transform(func)
```

Does `func` get called with Dask Series or pandas Series?

**Comment by OliverHofkens** [comment]
@TomAugspurger I copied the `groupby().apply()` behaviour, so the user-defined function receives pandas objects.

**Comment by TomAugspurger** [comment]
Hmm OK, I'll need to think on that. I think the most basic / useful user-defined transformation is something like a nomralize method that subtracts the group-wise mean

```python
df.groupby('col').transform(lambda series: series - series.mean())
```

I think we should aim to support that correctly. Do you see any hurdles to that? Will that always give the same output as pandas, or will partitioning get in the way?

**Comment by OliverHofkens** [comment]
Good point @TomAugspurger, I hadn't considered that. I think I test for your case in the tests I just added in 7d0afa39cb38612f70564a784b49296880a7fdd6. It seems to work fine for different partitioning layouts and both groupby on index and full-shuffle groupby, unless I'm missing something in my test.

**Comment by TomAugspurger** [comment]
Alright, thanks for checking. Seems like there's a related CI failure: https://travis-ci.org/dask/dask/jobs/578377303#L1056

File: docs/Makefile [hunk 4]
```diff
@@ -3,7 +3,7 @@
 
 # You can set these variables from the command line.
 SPHINXOPTS    =
-SPHINXBUILD   = sphinx-build
+SPHINXBUILD   = pipenv run sphinx-build
```
Comment (mrocklin) [inline docs/Makefile]: Why this change?

**Comment by mrocklin** [comment]
Any thoughts on the `test_groupby_transform_funcs[diff]` test failure?  It's odd that it failed in some test matrix elements, but not others.  I wonder if this is Pandas version dependent in some way.

File: docs/Makefile (hunk 4 above)
Comment (OliverHofkens) [reply docs/Makefile]: My bad, this shouldn't have been committed!

**Comment by OliverHofkens** [comment]
It does seem `pandas` related, the following `tox`file is able to reproduce the error: 

```ini
[tox]
envlist = {py35,py36,py37}-pandas{21,23,25}

[testenv]
extras =
    complete

deps =
    pytest
    pandas21: pandas==0.21.1
    pandas23: pandas==0.23.4
    pandas25: pandas==0.25.1

commands =
    pytest
```

I'll track down the exact problem and see what I can do.

**Comment by OliverHofkens** [comment]
It seems a lot of tests in `dask/dataframe/tests/test_groupby.py` are failing on lower pandas versions, even ones not related to my PR (on the master branch). I might be missing something else here. Looking into it!

**Comment by OliverHofkens** [comment]
I learned that `df.groupby().transform(series.diff)` returns something I didn't expect in earlier `pandas` versions (in both `pandas` and `dask`), seems the behavior changed in recent versions. I deleted the test in c42dcb75651665f04b124337568ce1da61ba7fd9

**Comment by mrocklin** [comment]
>  I deleted the test in c42dcb7

If the test is valuable then you might also consider skiping the test for lower versions, perhaps with `pytest.mark.skipif`

**Comment by mrocklin** [comment]
Merging this in 24 hours if there are no further comments

**Comment by OliverHofkens** [comment]
> > I deleted the test in [c42dcb7](https://github.com/dask/dask/commit/c42dcb75651665f04b124337568ce1da61ba7fd9)
> 
> If the test is valuable then you might also consider skiping the test for lower versions, perhaps with `pytest.mark.skipif`

I think the test with `pd.Series.rank` tests the same use case, I was just being redundant. 
Thanks @mrocklin, really excited to see this released!

**Comment by mrocklin** [comment]
Thanks @OliverHofkens .  This is in!

**Comment by CMCDragonkai** [comment]
This will be available on 2.4.0? When is that scheduled for?

**Comment by TomAugspurger** [comment]
Probably within a couple weeks.
